File: zigi-android/app/src/main/java/com/zigi/mobile/api/UssdApi.java
package com.zigi.mobile.api;

import com.zigi.mobile.Config;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

public class UssdApi {

    private static final String USSD_URL = Config.API_BASE + "/api/ussd";

    public static class UssdResult {
        public boolean ok;
        public boolean isFinal;
        public String display;
    }

    public static void callUssd(String sessionId, String phoneNumber, String text, ApiCallback<UssdResult> callback) {
        String formBody = encode("sessionId", sessionId) + "&" + encode("phoneNumber", phoneNumber) + "&" + encode("text", text);

        HttpUtil.postForm(USSD_URL, formBody, (ok, status, body, xUssdStatus) -> {
            UssdResult result = new UssdResult();
            result.ok = ok;
            result.display = ok ? body : ("Could not reach server. Is the backend running at " + Config.API_BASE + "?");
            result.isFinal = ok ? "END".equals(xUssdStatus) : true;
            callback.onResult(ok, result);
        });
    }

    private static String encode(String key, String value) {
        try {
            return key + "=" + URLEncoder.encode(value == null ? "" : value, StandardCharsets.UTF_8.name());
        } catch (Exception e) {
            return key + "=";
        }
    }
}

File: zigi_webapp/src/ussdApi.js
import { API_BASE, CLIENT_CHANNEL, getAuthToken } from './config.js'

const USSD_URL = `${API_BASE}/api/ussd`

export async function callUssd(sessionId, phoneNumber, text) {
  try {
    const headers = {
      "Content-Type": "application/x-www-form-urlencoded",
      "X-Client-Channel": CLIENT_CHANNEL,
    }
    const token = getAuthToken()
    if (token) headers["Authorization"] = `Bearer ${token}`

    const res = await fetch(USSD_URL, {
      method: "POST",
      headers,
      body: new URLSearchParams({ sessionId, phoneNumber, text }),
    })

    const display = await res.text()
    const isFinal = res.headers.get("X-Ussd-Status") === "END"
    return { ok: res.ok, isFinal, display }
  } catch (err) {
    return { ok: false, isFinal: true, display: "Could not reach server. Is backend running on :6157?" }
  }
}
